packages/block-library/src/navigation-link/index.php

```diff
@@ -369,3 +391,38 @@ function register_block_core_navigation_link() {
 	);
 }
 add_action( 'init', 'register_block_core_navigation_link' );
+// Register actions for all post types and taxonomies, to add variations when they are registered.
+// All post types/taxonomies registered before register_block_core_navigation_link, will be handled by that function.
+add_action( 'registered_post_type', 'register_block_core_navigation_link_post_type_variation', 10, 2 );
+add_action( 'registered_taxonomy', 'register_block_core_navigation_link_taxonomy_variation', 10, 3 );
```

Comment:
I think these two lines maybe causing the breaking PHPUNit tests for the PHP package sync process for WordPress 6.5

https://github.com/WordPress/wordpress-develop/pull/5922

Reply:
Maybe after #56952 was merged, we can just use the `variations_callback` function and don't need to hook into the `registered_post_type` and `registered_taxonomy` actions anymore. I can test locally. 

But do you have any idea as to why those hooks may cause phpunit timeouts?

Reply:
@getdave I've drafted up a PR which uses the now in core available `get_block_type_variations` filter instead of the hooks mentioned above. This should give us performance benefits, but also hopefully remove the problems with the unit tests you mentioned. 

https://github.com/WordPress/gutenberg/pull/58389

The PR is just a draft/exploration, because it will only work when that filter is available (which is core 6.5+). So it definitely has some BC pitfalls. Happy for any feedback or if you think, that this PR should just be reverted so gutenberg can successfully be merged into core.

Reply:
> This should give us performance benefits, but also hopefully remove the problems with the unit tests you mentioned.

Lets bring that in. I'll try it manually here.


> why those hooks may cause phpunit timeouts?

PHPUnit probably registers/unregisters post types hundreds of times. That _could_ be it? Still only a guess.

> The PR is just a draft/exploration, because it will only work when that filter is available (which is core 6.5+). So it definitely has some BC pitfalls. 

These Gutenberg changes will be for 6.5 so it should be ok. The only issue is that the Plugin should run with the last x2 versions of WP (I think) so we'd need to gate it based on the availability of the new `variations_callback ` API in Core. Can we do that?